### PR TITLE
Move Apple toolchain setup to apple_support

### PR DESCRIPTION
**Pull request by keith**
This moves the CC toolchain for building Apple platforms besides macOS to the apple_support repo https://github.com/bazelbuild/apple_support/pull/113

The default unix toolchain is now used if someone wants to build for macOS without the apple_support toolchain, but it doesn't handle as many platform specific features as the previous toolchain.

Fixes https://github.com/bazelbuild/bazel/issues/15041

### CODE REVIEW

**Comment by keith** [comment]
This is failing because rules_cc references the primary file we're deleting here

**Comment by keith** [comment]
Yea I definitely can, the problem will be doing that in lockstep, but I guess it's ok since it's breaking either way

**Comment by keith** [comment]
https://github.com/bazelbuild/rules_cc/pull/154

**Comment by keith** [comment]
I think the order of operations has to be:

1. remove from rules_cc
2. create rules_cc release, update it in bazel
3. remove here
4. merge into apple_support

**Comment by keith** [comment]
@katre ok I moved out all of the apple crosstool setup that I can, and moved it into https://github.com/bazelbuild/apple_support/pull/113/

one thing I'm unclear on currently is that I currently have the apple_support using `local_config_cc`, and falling back in the case the Apple toolchain doesn't apply. It seems ideally I would use `local_config_apple_cc` or something, but currently that would require we force a `--crosstool_top` argument that I don't think we want to. Based on https://github.com/grailbio/bazel-toolchain, this might change if we were using `--incompatible_enable_cc_toolchain_resolution`, but I believe that's still blocked on this detail https://github.com/bazelbuild/bazel/issues/7260#issuecomment-1255255323

**Comment by katre** [comment]
The right answer is for non-toolchains users to use `--crosstool_top=@rules_apple//whatever`, and for toolchains users to pick it up in the WORKSPACE.

That sequencing looks right but we should confirm with @comius.

**Comment by keith** [comment]
I wonder if adding that in our transition would be enough instead of requiring that change on folks before we transition to the workspace way

**Comment by keith** [comment]
it seems like if we could change the default of `apple_crosstool_top` to the new toolchain _when it exists_ users wouldn't have to know about this. we could do that always in rules_apple, but that also is used in a transition in this repo, and defaults to the default CC toolchain

**Comment by keith** [comment]
I think this is pretty close to ready, but I can't make CI green until we have a new release of rules_cc, it seems too difficult to correctly override all the tests cases for it at the moment.

File: tools/cpp/unix_cc_toolchain_config.bzl
Comment (brentleyjones) [inline tools/cpp/unix_cc_toolchain_config.bzl:1394]: nit: should be above `provides`?

File: tools/cpp/unix_cc_configure.bzl
Comment (keith) [inline tools/cpp/unix_cc_configure.bzl:492]: This is required because now this toolchain can be used when the user has Xcode + the CLT installed, in which case the Xcode paths are automatically collected, but this SDK can still be used.

File: tools/cpp/unix_cc_toolchain_config.bzl
Comment (keith) [inline tools/cpp/unix_cc_toolchain_config.bzl:300]: This is required to get reproducible archives / links on macOS, it shouldn't hurt applying to other platforms

File: tools/cpp/unix_cc_toolchain_config.bzl
Comment (keith) [inline tools/cpp/unix_cc_toolchain_config.bzl:1302]: this is required because when you have Xcode + the CLT installed, it seems that the minimum os version behavior changes somehow, without explicitly passing this I saw it defaulting to the oldest version. This also means `--macos_minimum_os` works for C* binaries

File: tools/cpp/unix_cc_toolchain_config.bzl
Comment (keith) [inline tools/cpp/unix_cc_toolchain_config.bzl:1355]: This was not supposed to be here, but was just not getting exercised. see https://github.com/bazelbuild/bazel/issues/12439

File: tools/cpp/unix_cc_toolchain_config.bzl
Comment (keith) [inline tools/cpp/unix_cc_toolchain_config.bzl:1370]: This was not supposed to be here and not in the default Xcode toolchain

File: MODULE.bazel
Comment (katre) [inline MODULE.bazel:14]: Presumably this needs to be a released rules_cc at some point, just flagging so it's not forgotten.

File: src/main/starlark/builtins_bzl/common/objc/objc_library.bzl
Comment (katre) [inline src/main/starlark/builtins_bzl/common/objc/objc_library.bzl:50]: Nice!

File: src/test/java/com/google/devtools/build/lib/pkgcache/BUILD
Comment (katre) [inline src/test/java/com/google/devtools/build/lib/pkgcache/BUILD:281]: Should this be removed, or is the test going to be fixed?
Prefer tagging as "noci" if the test doesn't pass on CI.

File: src/test/py/bazel/bzlmod/bazel_module_test.py
Comment (katre) [inline src/test/py/bazel/bzlmod/bazel_module_test.py:26]: Is this still needed after rules_cc has your changes?

File: src/test/shell/bazel/BUILD
Comment (katre) [inline src/test/shell/bazel/BUILD:372]: Same comment about using "noci", and whether this test can be re-enabled.

File: src/test/shell/bazel/BUILD
Comment (katre) [inline src/test/shell/bazel/BUILD:1058]: And here.

File: src/test/shell/bazel/apple/bazel_apple_test.sh
Comment (katre) [inline src/test/shell/bazel/apple/bazel_apple_test.sh:43]: Are these tests moving to apple_support with the toolchain, or are they no longer needed at all?

File: tools/cpp/unix_cc_configure.bzl
Comment (katre) [reply tools/cpp/unix_cc_configure.bzl:492]: Can you add that as a comment in the code?

File: tools/cpp/unix_cc_toolchain_config.bzl
Comment (katre) [reply tools/cpp/unix_cc_toolchain_config.bzl:1302]: Definitely make this a comment.

File: MODULE.bazel
Comment (keith) [reply MODULE.bazel:14]: yep! Hoping that @oquenchil can help with a release after https://github.com/bazelbuild/rules_cc/pull/154 is merged (which hopefully is in progress)

File: src/test/py/bazel/bzlmod/bazel_module_test.py
Comment (keith) [reply src/test/py/bazel/bzlmod/bazel_module_test.py:26]: nope, just need that update

File: src/test/shell/bazel/BUILD
Comment (keith) [reply src/test/shell/bazel/BUILD:372]: nope, just need that update

Comment (keith) [reply src/test/shell/bazel/BUILD:372]: I'll just leave these as broken since they shouldn't ever land this way

File: src/test/shell/bazel/BUILD
Comment (keith) [reply src/test/shell/bazel/BUILD:1058]: I'll just leave these as broken since they shouldn't ever land this way

File: src/test/shell/bazel/apple/bazel_apple_test.sh
Comment (keith) [reply src/test/shell/bazel/apple/bazel_apple_test.sh:43]: these all moved to apple_support. anything that relied on the real toolchain and non macOS basic features has to move there. there are still many tests here using the mock toolchain for coverage

**Comment by keith** [comment]
Threw the comments in here, didn't change the broken tests since we shouldn't land until we have the rules_cc  update here, at which point I'll remove them from this diff

File: src/test/java/com/google/devtools/build/lib/packages/util/mock/osx_cc_toolchain_config.bzl
Comment (brentleyjones) [inline src/test/java/com/google/devtools/build/lib/packages/util/mock/osx_cc_toolchain_config.bzl:2333]: Why are these needed?

File: tools/cpp/unix_cc_toolchain_config.bzl
Comment (brentleyjones) [reply tools/cpp/unix_cc_toolchain_config.bzl:1370]: Probably too much effort, but should we land these changes first?

File: src/test/java/com/google/devtools/build/lib/packages/util/mock/osx_cc_toolchain_config.bzl
Comment (keith) [reply src/test/java/com/google/devtools/build/lib/packages/util/mock/osx_cc_toolchain_config.bzl:2333]: this is what objc_library now uses to tell if it's supported to build objc

**Review by oquenchil — APPROVED** [review]

**Comment by brentleyjones** [comment]
@oquenchil This can't be merged yet, it has commented out stuff waiting on the new 0.0.5 rules_cc release. Can you change the labels to prevent it accidentally landing early?

**Comment by keith** [comment]
Yep good idea, just reverted all the test workarounds so this will be red and blocked on https://github.com/bazelbuild/rules_cc/issues/174 until we update bazel with that new version

**Comment by sgowroji** [comment]
Hi @keith, Above PR is awaiting to merge. Could you please fix the above branch conflicts and  Pending [changes](https://github.com/bazelbuild/bazel/pull/16619#pullrequestreview-1264017133). Thanks!

**Comment by brentleyjones** [comment]
@sgowroji still blocked on https://github.com/bazelbuild/rules_cc/issues/174. The red statuses are expected until that.

**Comment by brentleyjones** [comment]
rules_cc 0.0.5 has been released. Keith is OOO for another week, so no progress will be made on this PR until at least then.

**Comment by keith** [comment]
We're all green here 🎉 although we should probably merge https://github.com/bazelbuild/bazel/pull/17475 first to simplify reverts if needed.

@oquenchil @katre could you run the downstream pipeline job with this change? Also is there anything else on your mind that we should do before merging?

**Comment by meteorcloudy** [comment]
https://buildkite.com/bazel/bazel-at-head-plus-downstream/builds/2873

**Comment by keith** [comment]
thanks @meteorcloudy can you run it one more time with the fixes I pushed for those? (note that many of those issues were not introduced by this PR)

**Comment by oquenchil** [comment]
New one here: https://buildkite.com/bazel/bazel-at-head-plus-downstream/builds/2876

**Comment by keith** [comment]
@oquenchil @katre I submitted PRs for all the downstream breakages from this, can we talk about a merge plan?

**Comment by oquenchil** [comment]
Thanks Keith. What are the reasons for a specific plan instead of just merging the PR normally?

**Comment by keith** [comment]
I guess I just wanted to hear from you all if you have any specific concerns. I imagine there will be some straggler issues that come up and I'm happy to field them. Assuming they're not catastrophic I think it would be nice if we could avoid reverting since it's a conflict-prone change

**Comment by keith** [comment]
but if you all have no other concerns then this is ready to go!

**Comment by oquenchil** [comment]
Still working on merging this.

**Comment by googlewalt** [comment]
I have a concern about migrating the OSS apple crosstool from bazel to a separate repository.  In the past I have had to make changes to the OSS apple crosstool so that it works with changes I'm making to bazel.  How do you envision this being handled in the future, if such a need arises?  Are the apple_support repo owners prepared to handle these changes, and any issues that may arise with keeping the toolchain compatible with multiple versions of bazel.

A couple options that would lessen/eliminate my concern:

- keep a copy of the apple crosstool in bazel, so that me and other bazel developers can update the crosstool in sync with bazel changes, that apple_support can import as needed.

- keep the crosstool in bazel, but proceed with the rest of the proposed migration.

**Comment by keith** [comment]
Given the historical examples I don't think the rate or complexity of changes that you're referring to will be a problem for this living in another repo. I think depending on the changes you can submit PRs, or we could adapt them ourselves.

I don't think duplicating or leaving any of the moved files here achieves the goals we set out to solve. Duplicating the toolchain would increase the burden on the rules maintainers, and both of those options would continue to block supporting platforms with this toolchain in the future.

I also wonder how these concerns apply to other custom toolchains folks are using as well? For example if we duplicated the toolchain, the downstream pipeline would still break if bazel made breaking changes that also required toolchain changes, similar to bazel-toolchain or other popular CC toolchains that aren't part of this repo.

**Comment by googlewalt** [comment]
We see a bunch of failures: https://buildkite.com/bazel/google-bazel-presubmit/builds/64142. We're fixing it in the internal import patch, but I want to check whether this should have been caught by bazel presubmits earlier?

**Comment by keith** [comment]
i imagine something broke since i rebased last?

**Comment by keith** [comment]
let me know if you want me to fix on this branch!

**Comment by keith** [comment]
Thanks all! @googlewalt just so I know for the future, are the files you ended up keeping libtool.sh / make_hashed_objlist.py still there because they're used in those paths internally? (since i don't see any external references)

**Comment by meteorcloudy** [comment]
image
https://buildkite.com/bazel/bazel-bazel/builds?branch=master&page=3
https://buildkite.com/bazel/bazel-bazel/builds/22421#0186ca61-268c-4135-b44d-f1123f11d11c

It looks like this PR is causing a lot of tests to time out in Bazel's postsubmit. @keith Can you take a look?

**Comment by meteorcloudy** [comment]
I guess we need to increase the test size of those timing out tests, e.g. https://cs.opensource.google/bazel/bazel/+/master:src/test/shell/bazel/BUILD;l=937

But this does indicate that there is a potential performance regression?

**Comment by googlewalt** [comment]
> Thanks all! @googlewalt just so I know for the future, are the files you ended up keeping libtool.sh / make_hashed_objlist.py still there because they're used in those paths internally? (since i don't see any external references)

Yes.  I suspect we should get rid of that dependency but left the cleanup for another day.

**Comment by keith** [comment]
@meteorcloudy I'll look at this today

**Comment by keith** [comment]
https://github.com/bazelbuild/bazel/pull/17763

**Comment by meteorcloudy** [comment]
This change was rollbacked at https://github.com/bazelbuild/bazel/commit/a50cca561fe0a77c3750101e505171ddcc16db72, @zhengwei143 can you provide more details about what was broken?

**Comment by brentleyjones** [comment]
I believe @googlewalt is working on a roll forward.

**Comment by keith** [comment]
@googlewalt how's the forward fix looking?

**Comment by googlewalt** [comment]
I just fixed one small remaining issue.  Awaiting review @oquenchil.

**Comment by googlewalt** [comment]
Submitted as https://github.com/bazelbuild/bazel/commit/699e40373f95e42390a85f29dfa1098636336103.

**Comment by keith** [comment]
Thanks!

**Pull request closed without merging**
